core/src/main/scala/nl/vroste/zio/kinesis/client/zionative/Consumer.scala

```diff
@@ -426,6 +426,95 @@ object Consumer {
+   * @param deserializer
+   *   Record deserializer
+   * @param workerIdentifier
+   *   Identifier of this worker, used for lease coordination. Must be unique for each worker
+   * @param fetchMode
+   *   How to fetch records: Polling or EnhancedFanOut, including config parameters
+   * @param leaseCoordinationSettings
+   *   Config parameters for lease coordination
+   * @param initialPosition
+   *   When no checkpoint exists yet for a shard, start processing from this position
+   * @param emitDiagnostic
+   *   Function that is called for events happening in the Consumer. For diagnostics / metrics
+   * @param shardAssignmentStrategy
+   *   How to assign shards to this worker
+   * @param checkpointBatchSize
+   *   Maximum number of records before checkpointing
+   * @param checkpointDuration
+   *   Maximum interval before checkpointing
+   * @param checkpointRetrySchedule
+   *   Schedule for retrying checkpointing in case of failure
+   * @param processorParallelism
+   *   Number of fibers to process records in parallel per shard
+   * @param recordProcessor
+   *   A function for processing a `Record[T]`
+   * @return
+   *   A ZIO that completes with Unit when record processing is stopped or fails when the consumer stream fails
+   */
+  def consumePartitionedWith[R, RC, T](
+    streamIdentifier: StreamIdentifier,
+    applicationName: String,
+    deserializer: Deserializer[R, T],
+    workerIdentifier: String = "worker1",
+    fetchMode: FetchMode = FetchMode.Polling(),
+    leaseCoordinationSettings: LeaseCoordinationSettings = LeaseCoordinationSettings(),
+    initialPosition: InitialPosition = InitialPosition.TrimHorizon,
+    emitDiagnostic: DiagnosticEvent => UIO[Unit] = _ => ZIO.unit,
+    shardAssignmentStrategy: ShardAssignmentStrategy = ShardAssignmentStrategy.balanced(),
+    checkpointBatchSize: Long = 200,
+    checkpointDuration: Duration = 5.minutes,
+    checkpointRetrySchedule: Schedule[R, Throwable, Any] =
+      Util.exponentialBackoff(min = 1.second, max = 1.minute, maxRecurs = Some(5)),
+    processorParallelism: Int = 1
+  )(
+    recordProcessor: Record[T] => RIO[RC, Unit]
+  ): ZIO[
+    R with RC with Kinesis with LeaseRepository,
+    Throwable,
+    Unit
+  ] =
+    shardedStream(
+      streamIdentifier,
+      applicationName,
+      deserializer,
+      workerIdentifier,
+      fetchMode,
+      leaseCoordinationSettings,
+      initialPosition,
+      emitDiagnostic,
+      shardAssignmentStrategy
+    ).flatMapPar(Int.MaxValue) { case (_, shardStream, checkpointer) =>
+      ZStream.fromZIO(LockingCheckpointer.make[T](checkpointer, checkpointBatchSize, checkpointDuration)).flatMap {
+        lockingCheckpointer =>
+          shardStream
+            .tap(lockingCheckpointer.lock)
+            // Use mapZIOPar even when parallelism is 1 to ensure that locking happens on a seperate fiber.
+            .mapZIOPar(processorParallelism)(r => recordProcessor(r).as(r))
```

Comment:
With this alternative, the lock and release bit becomes a bit more intuitive (in my opinion). Or would that not work because of the separate fiber (can you explain why that is necessary)?

```scala
              ZIO
                .acquireReleaseWith(lockingCheckpointer.lock(r))(_ => lockingCheckpointer.stage(r))(_ =>
                  recordProcessor(r)
                )
```

Reply:
The reason to have it this way is that lockingCheckpointer.lock will block the fiber the calls it. If we call it on the same fiber that is actually processing the record, we are not going to make any progress.

The way that it works right now is we have a single fiber locking (and then getting blocked when it cannot lock the record) and submitting the records of the internal buffer of mapZIOPar.
We then have processorParallelism fibers (which are not affected by the blocking due to locking) that take records from the internal buffer and process + potentially unlock.

Reply:
Ah I see, you want to be able to process more than 1 record per shard at the same time (but not more than 1 per partition key). That indeed won't work with my suggestion. I missed that part, thinking the only goal here was to ensure that there's never more than 1 uncheckpointed record per partition key. 

I'm mostly using kafka nowadays, where your parallelism is typically limited by the number of partitions which you choose to be some number that divides into several smaller numbers. Is it not common in the kinesis world to choose a number of shards that is larger than the parallelism you desire? And if so is the parallelism needed?

Reply:
Yeah, you would normally do this using the number of shards. TBH the parallelism mostly shows up due to an implementation detail of wanting to run the processing on separate fibers.
As you say, it might be useful sometimes though so I decided to expose it as a setting.